src/reconstruction/community.jl

```diff
@@ -0,0 +1,406 @@
+    end
+
+    # # build diagonal environmental exchanges
+    for i = 1:length(exchange_rxn_ids)
+        I[kstart] = n_metabolites_metabolic + i
+        J[kstart] = n_reactions_metabolic + i
+        V[kstart] = -1.0
+        kstart += 1
+    end
+
+    if add_biomass_objective
+        n_before_biomass_row = n_metabolites_metabolic + length(exchange_met_ids)
+        for i = 1:length(models)
+            biomass_ind = first(indexin([biomass_ids[i]], reactions(models[i])))
+            I[kstart] = i + n_before_biomass_row
+            J[kstart] = biomass_ind + reaction_offset[i]
+            V[kstart] = 1.0
+            kstart += 1
+        end
+    end
+
+    S = sparse(
+        I[1:kstart-1],
+        J[1:kstart-1],
+        V[1:kstart-1],
+        n_metabolites_total,
+        n_reactions_total,
+    ) # could be that some microbes don't have all the exchanges, hence kstart-1
+
+    reaction_cumsum = cumsum(reaction_lengths)
+    metabolite_cumsum = cumsum(metabolite_lengths)
+    for i = 1:length(models)
+        species = isempty(model_names) ? "species_$(i)" : model_names[i]
+        tlbs, tubs = bounds(models[i])
+        lbs[reaction_offset[i]+1:reaction_cumsum[i]] .= tlbs
+        ubs[reaction_offset[i]+1:reaction_cumsum[i]] .= tubs
+        rxns[reaction_offset[i]+1:reaction_cumsum[i]] =
+            "$(species)_" .* reactions(models[i])
+        mets[metabolite_offset[i]+1:metabolite_cumsum[i]] =
+            "$(species)_" .* metabolites(models[i])
+    end
+    mets[metabolite_cumsum[end]+1:metabolite_cumsum[end]+length(exchange_met_ids)] .=
+        exchange_met_ids
+    rxns[reaction_cumsum[end]+1:reaction_cumsum[end]+length(exchange_rxn_ids)] .=
+        exchange_rxn_ids
+
+    if add_biomass_objective
+        rxns[end] = "community_biomass"
+        for i = 1:length(models)
+            species = isempty(model_names) ? "species_$(i)" : model_names[i]
+            mets[end-length(biomass_ids)+i] = "$(species)_" .* biomass_ids[i]
+        end
+    end
+
+    return CoreModel(S, spzeros(size(S, 1)), spzeros(size(S, 2)), lbs, ubs, rxns, mets)
+end
+
+"""
+    boundary_reactions_metabolites(model::MetabolicModel)::Tuple{Vector{String}, Vector{String}}
+
+Return `boundary_rxns` and `boundary_mets` given `model`. Boundary reactions
+are reactions with a single stoichiometric coefficient in a column in the
+stoichiometric matrix, and boundary metabolites are the corresponding row/metabolite
+for that column.
+"""
+function boundary_reactions_metabolites(
+    model::MetabolicModel,
+)::Tuple{Vector{String},Vector{String}}
+    boundary_mets = String[]
+    boundary_rxns = String[]
+    S = stoichiometry(model)
+    rxns = reactions(model)
+    mets = metabolites(model)
+    for i = 1:size(S, 2)
+        n = nnz(S[:, i])
+        if n == 1
+            j, _ = findnz(S[:, i])
+            push!(boundary_mets, mets[first(j)])
+            push!(boundary_rxns, rxns[i])
+        end
+    end
+    return boundary_rxns, boundary_mets
```

Comment:
perhaps better return a named tuple here so that people don't need to bind this explicitly or pick it up using [1] [2]

Reply:
or maybe I should remove this function? with `looks_like` this isn't necessary anymore... what do you think?

Reply:
We could have another `looks_like` for boundaries based on stoichiometry?

Reply:
I don't think it should be called `looks_like` in that case since you can directly test if it is a boundary reaction (they only have one metabolite?).  That might be a good function to include somewhere in `src/base/util` though. For now I have removed this function